### DISCUSSION TITLE
Lesson 12: Test reverts fine but Health Factor value doesn't change.

### DISCUSSION THREAD

**Discussion by itznishant** [Cyfrin/foundry-full-course-cu#679] [Q&A]
Lesson 12: I am trying to run this foundry test ```testShouldRevertIfHealthFactorBreaksOnMintingDsc``` on mintDsc function which should revert since USER is trying to mint DSC tokens without depositing any collateral.

<img width="665" alt="__test__" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/a09514c4-b333-4f26-9755-ffe1bb28c9f3">

It reverts correctly, but the health factor of user doesn't change as seen below.
<img width="863" alt="test_output1" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/c0af4fc5-4ffb-4dc8-9945-f3e80838ff6a">

I am unable to understand what's happening here. Is this related to the health factor bug Patrick mentions in the video ?

The getter function to retrieve the health factor for the USER is below.

```
    function getHealthFactor(address user) public view returns (uint256) {
        return _healthFactor(user);
    }
}
```

TIA for your help!

Nishant

**Comment by 0xusmanf** [comment]
The health factor of the user will remain unaffected unless there are modifications to his collateral or DSC balance. Your test is performing correctly, as the health factor should remain unchanged before and after the mintDSC function call. This is because the function didn't alter the user's DSC balance due to a reverted transaction, resulting in no change to the contract state.

**Reply by itznishant** [reply]
Don't think the test is performing correctly somewhere. 
The function ```revertifHealthFactorBreaks``` below has a conditional revert based on the healthfactor value > MIN_HEALTH_FACTOR.
But, the health factor shown is some very large value, greater than MIN_HEALTH_FACTOR so why's this conditon being triggered, when it shouldn't?

<img width="664" alt="image" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/896c7cb4-0c9b-4bab-8f8c-77d30cd2579f">

Shouldn't the userhealthfactor = 0 . Because collateralValueInUsd would be 0 in this case, right?

Really need to understand what's going on here, to proceed further.

**Reply by 0xusmanf** [reply]
> The function revertifHealthFactorBreaks below has a conditional revert based on the healthfactor value > MIN_HEALTH_FACTOR.
> But, the health factor shown is some very large value, greater than MIN_HEALTH_FACTOR so why's this conditon being triggered, when it shouldn't?

In case where you are calling `dscEngine.getHealthFactor(USER)` you are getting a very large value in return due to the following condition in `_calculateHealthFactor` function `if (totalDscMinted == 0) return type(uint256).max;`. So in your case USER's DSC balance is 0, that's why this if condition trigers and returns `type(uint256).max;` which is `115792089237316195423570985008687907853269984665640564039457584007913129639935`.

but when you try to call `dscEngine.mintDsc(DSC_TOKRNS_TO_MINT)` it reverts because it updates USER's DSC balance with `DSC_TOKRNS_TO_MINT` and call `_calculateHealthFactor`, so in this case the following code will execute

```
        uint256 collateralAdjustedForThreshold = (collateralValueInUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
        return (collateralAdjustedForThreshold * 1e18) / totalDscMinted;
```

which result in returning 0 as `collateralValueInUsd` will be 0, hence following if condition will be true and it will revert undoing any state changes i.e it will set USER's DSC balance to its initial state, which is 0.

```
if (userHealthFactor < MIN_HEALTH_FACTOR) {
            revert DSCEngine__BreaksHealthFactor(userHealthFactor);
}
```

**Reply by itznishant** [reply]
okay I got the rationale for that large number for **before** case where totalDscMinted = 0.
But based on your explanation, the _calculateHealthFactor and thereby _healthFactor should return 0 and the **after** console log should display healthFactor as  0 right!

**Reply by 0xusmanf** [reply]
> based on your explanation, the _calculateHealthFactor and thereby _healthFactor should return 0 and the after console log should display healthFactor as 0 right!

Can you please explain your thought process line by line why you expect the health factor to be zero in **after console log**. Maybe I am missing something here.

**Comment by theirrationalone** [comment]
This might help you [discussion](https://github.com/Cyfrin/foundry-full-course-f23/discussions/523#discussioncomment-6630649) 

Also read this too: [comment](https://github.com/Cyfrin/foundry-full-course-f23/discussions/403#discussioncomment-6706197)

**Reply by itznishant** [reply]
> The health factor of the user will remain unaffected unless there are modifications to his collateral or DSC balance. Your test is performing correctly, as the health factor should remain unchanged before and after the mintDSC function call. This is because the function didn't alter the user's DSC balance due to a reverted transaction, resulting in no change to the contract state.



> This might help you [discussion](https://github.com/Cyfrin/foundry-full-course-f23/discussions/523#discussioncomment-6630649)
> 
> Also read this too: [comment](https://github.com/Cyfrin/foundry-full-course-f23/discussions/403#discussioncomment-6706197)

Thanks for the reply. It seems though that these are realted to the liquidation functionality and some collateral deposited?
In my test case, user hasn't deposited any collateral and so the mintdsc should revert if they try to mint any amonut of dsc without any backing.

I have added additonal details to @usmanfarooq91 answer, can you check and help with that.
Thanks!

**Reply by theirrationalone** [reply]
Thank You @itznishant ! For Finding this Great **BUG** (Logical in Sense), Although I made a PULL REQUEST on it with my mitigation logic. Wait until then.

**Comment by itznishant** [comment]
> 



> > The function revertifHealthFactorBreaks below has a conditional revert based on the healthfactor value > MIN_HEALTH_FACTOR.
> > But, the health factor shown is some very large value, greater than MIN_HEALTH_FACTOR so why's this conditon being triggered, when it shouldn't?
> 
> In case where you are calling `dscEngine.getHealthFactor(USER)` you are getting a very large value in return due to the following condition in `_calculateHealthFactor` function `if (totalDscMinted == 0) return type(uint256).max;`. So in your case USER's DSC balance is 0, that's why this if condition trigers and returns `type(uint256).max;` which is `115792089237316195423570985008687907853269984665640564039457584007913129639935`.
> 
> but when you try to call `dscEngine.mintDsc(DSC_TOKRNS_TO_MINT)` it reverts because it updates USER's DSC balance with `DSC_TOKRNS_TO_MINT` and call `_calculateHealthFactor`, so in this case the following code will execute
> 
> ```
>         uint256 collateralAdjustedForThreshold = (collateralValueInUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
>         return (collateralAdjustedForThreshold * 1e18) / totalDscMinted;
> ```
> 
> which result in returning 0 as `collateralValueInUsd` will be 0, hence following if condition will be true and it will revert undoing any state changes i.e it will set USER's DSC balance to its initial state, which is 0.
> 
> ```
> if (userHealthFactor < MIN_HEALTH_FACTOR) {
>             revert DSCEngine__BreaksHealthFactor(userHealthFactor);
> }
> ```



> Thank You @itznishant ! For Finding this Great **BUG** (Logical in Sense), Although I made a PULL REQUEST on it with my mitigation logic. Wait until then.

Hmm, I think you should explain the issue in detail here first for all of us to understand.

Is this BUG related to the fact that the function does NOT check if collateral is even deposited by user before checking for totalDscMinted values?

Or that the user healthFactor doesn't update correctly for a reverted case, it still holds some stale value  ?

I should probably raise the PR since I posted this issue here! You helped a lot of people here already, let me have at least 1 PR contribution on this course. Going through tough times, should cheer me up.

**Reply by theirrationalone** [reply]
Why not buddy, you can make my words yours, Actually all words already yours i just optimized them. 😉

Heyy, @usmanfarooq91 ! You should read this... 

The Bug is, If an user isn't participating at all Or if he hasn't any collateral amount deposited into account. So In fact, His/Her Health Factor must be Zero. 

Now @itznishant !, Buddy it's your turn to explain all of this, after then you can mark your words as the answer.

**Reply by 0xusmanf** [reply]
> Or that the user healthFactor doesn't update correctly for a reverted case, it still holds some stale value ?

I think you are misunderstanding how a revert works, in case of a revert, any changes made within that transaction are effectively undone. Consequently, there should be no change to the user's health factor.

**Reply by 0xusmanf** [reply]
> Heyy, @usmanfarooq91 ! You should read this...
> 
> The Bug is, If an user isn't participating at all Or if he hasn't any collateral amount deposited into account. So In fact, His/Her Health Factor must be Zero.

`type(uint256).max` value returned indicates that either there is **0 DSC minted and 0 collateral deposited** or **0 DSC minted and X amount of collateral deposited**, there is no way where there is a possibility of **X amount of DSC minted and 0 collateral deposited**, so, in my opinion, it's not a bug and it doesn't affect the protocol in any way. I am open to being persuaded to a different viewpoint and would welcome it.

**Reply by theirrationalone** [reply]
Ahhaaa, You're so close to get it.... That's not true for 0 collateral amount... Now you're smart enough 😉

**Reply by 0xusmanf** [reply]
I know what you are referring to, I have seen your PR 😉, but again what difference does it make?

**Reply by theirrationalone** [reply]
With this, I think Protocol won't need to spend its usage, electricity & energy, to perform illogical txs... as well as protocol won't need to calculate that `max.` value without any need. IMO, It would also prevent malicious performances by stopping attackers at the Gate. "Protocol: Heyy You, stop here you're not participating, please subscribe me Or go away.., I don't mean you, I'm not going to blacklist you just go away Or participate if you wish then please put/deposit some collateral Or just Get Lost 😂😉".

@usmanfarooq91 what's your opinion?? would it be verbose??, crossing my fingers, wish Patrick Approves my PR 😅😅😅

**Reply by 0xusmanf** [reply]
I wish you good luck on your PR but I think This change will not make sense to the students as it is not mentioned in the tutorial and cause confusion.

**Reply by theirrationalone** [reply]
@usmanfarooq91 **THANK YOU** for the Wish ♥,

Crossing My Fingers 🤞!, Should be the First thing for students to keep in mind right before diving into the Protocol Implementation.

Credits To: @itznishant !, He is the Founder of this **Discovery** | **BUG** | **VERBOSE IMPLEMENTATION but Logical**! 😉
